### PR TITLE
thrift: Reject router config if cluster/weighted clusters/mirror cluster doesn't exist for static route config provider

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:
Expose `ServerFactoryContext` to ConfigTraitsImpl::createConfig so the underlying `RouteEntryImpl` is able to validate its clusters.

Currently HTTP validates the clusters for static route config by default but doesn't validate the clusters for RDS.
Moreover, optional field `validate_clusters` changes the default behavior.

In this diff, thrift route config provider manager do the same default behavior for cluster validation.
That is, validate the clusters for static route config by default but *not* validate the clusters for TRDS.

Optional field `validate_clusters` is coming soon.

Risk Level: low
Testing: unit test
[Optional Fixes #Issue] #20559

### CODE REVIEW

**Comment by JuniorHsu** [comment]
Before I fixed all the interesting tests and add the appropriate test coverage. 

Like http config setup, we don't validate the cluster for rds. Instead, only validate the clusters in static config.

Which introduce two interfaces I'm hesitating to determine:

1. 
```
virtual ConfigConstSharedPtr createConfig(const Protobuf::Message& rc,
                                            Server::Configuration::ServerFactoryContext& context,
                                            bool validate_clusters_default) const PURE;
```

I'm not sure if we need to have the flexibility the change the `validate_clusters_default` while `createConfig`
Or once we decide the configTraits, the rule is fixed.

Currently I keep the flexibility since reverting interface is expensive .
2.
```
RouteMatcher::RouteMatcher(
    const envoy::extensions::filters::network::thrift_proxy::v3::RouteConfiguration& config,
    Server::Configuration::ServerFactoryContext& factory_context, bool validate_clusters)
```

Same thing, for thrift RouteMatcher, we could pass the limited information `absl::optional<Upstream::ClusterManager::ClusterInfoMaps>` instead of while context.

It's not an interface, we might want the minimal exposure principal?

cc @zuercher @rgs1 @tkovacs-2

**Comment by JuniorHsu** [comment]
After I take another look, 1. is not an issue given the config traits if common/rds doesn't know if it should create a config with validation. Unless the traits take another parameters.

I'll go ahead with 2. to minimal exposure and fix the tests.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:227]: not needed:

https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/thrift_proxy/v3/route.proto#L102

**Review by rgs1 — COMMENTED** [review]
Generally LGTM, just one nit

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:227]: Great catch. Will fix.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (tkovacs-2) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1953]: I think this assertion could be done in the UnknownCluster2.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (tkovacs-2) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1978]: Do we really need this test case?

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (tkovacs-2) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:2021]: Do we really need this test case?
If yes, then the config initialization is quite copy-paste. It could be extracted to a function like 
`initRouteConfig(..., const std::vector<std::string>& cluster_names, std::vector<int>& weights>)` or something.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1978]: it's kinda random tests, but I could merge 2 and 3 to make it more elegant

**Comment by tkovacs-2** [comment]
> I'm not sure if we need to have the flexibility the change the `validate_clusters_default` while `createConfig`. Or once we decide the configTraits, the rule is fixed.

Since it is not protocol specific, I think the `validate_clusters_default` is ok here. 

> It's not an interface, we might want the minimal exposure principal?

I also vote for minimal exposure.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1953]: You mean this is covered by the next test case? That makes sense to remove this given the next test case is not complicated.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:2021]: Again it's just random thought, the weight shouldn't matter. I'll merge these two cases.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (tkovacs-2) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc]: I mean you can move the `EXPECT_THROW_WITH_REGEX(....` into the next case (and delete this case).
For that call the same config is ok what is prepared in the next case.

**Comment by tkovacs-2** [comment]
Don't you add flag for the validation into the config proto?
I'm not sure but in that case if the listener and the routing is in the bootstrap config but the clusters are through xDS then the validation fails at startup.

**Comment by JuniorHsu** [comment]
> Don't you add flag for the validation into the config proto? I'm not sure but in that case if the listener and the routing is in the bootstrap config but the clusters are through xDS then the validation fails at startup.

No. Currently follow what the http did: only validate in static provider. No config proto is able to change the behavior at the moment.

As for your scenario, envoy will retry after the cluster arrives after it fails at startup without cluster information. Not perfect but it works.

**Comment by JuniorHsu** [comment]
```
Writing fixes to clang-tidy-fixes.yaml ...
clang-tidy check failed, potentially fixed by clang-apply-replacements:
Diagnostics:
- BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
  DiagnosticMessage:
    FileOffset: 298
    FilePath: ./source/common/common/logger.h
    Message: '''source/common/common/logger_impl.h'' file not found'
    Replacements: []
  DiagnosticName: clang-diagnostic-error
  Level: Error
  Ranges:
  - FileOffset: 298
    FilePath: ./source/common/common/logger.h
    Length: 36
```
For the test failure in "envoy-presubmit (check linux_x64 clang_tidy)" , I guess it's unrelated since I didn't touch nor include the logger.

**Comment by JuniorHsu** [comment]
Note:  We might want `validate_clusters` in proto at some point like http.
https://github.com/envoyproxy/envoy/blob/a308789897793b111ee65715c5ddb4a58b775083/api/envoy/config/route/v3/route.proto#L110

**Review by rgs1 — COMMENTED** [review]
This looks reasonable to me, thanks. 

@htuch do you have some cycles to look at the common RDS code?

**Comment by tkovacs-2** [comment]
> Note: We might want `validate_clusters` in proto at some point like http.

I would set the default to false even for of static case and add the flag to the proto to control the validation.
The current change is not fully backward compatible: If someone has a dummy/legacy rule referring to an unknown cluster but never matches, this change breaks a previously working system.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (tkovacs-2) [inline test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1956]: nit: if you want, you can add back the assert for the case when the cluster list is empty.

File: envoy/rds/config_traits.h
Comment (htuch) [inline envoy/rds/config_traits.h:60]: Please document the new params in the Doxygen-style comments above.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (htuch) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:60]: Isn't this the same logic as in `source/common/router/conifg_impl.cc`?

Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:60]: Yes, as line 51 said, it's similar validation procedure with `Envoy::Router::RouteEntryImplBase::validateCluster`
However, they are not with the same interface (as I need to adjust some method name). I have no idea how much effort reusing the logic is.

File: envoy/rds/config_traits.h
Comment (JuniorHsu) [reply envoy/rds/config_traits.h:64]: done. thanks.

File: test/extensions/filters/network/thrift_proxy/conn_manager_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/network/thrift_proxy/conn_manager_test.cc:1954]: done. thanks!

File: source/common/rds/route_config_update_receiver_impl.cc
Comment (htuch) [inline source/common/rds/route_config_update_receiver_impl.cc:19]: There is a `RouteConfiguration` option to do this validation IIRC,  do you have the same in TRDS? Or are you just relying on always validating?

Comment (JuniorHsu) [reply source/common/rds/route_config_update_receiver_impl.cc:19]: HTTP has an option to change.
https://github.com/envoyproxy/envoy/blob/a308789897793b111ee65715c5ddb4a58b775083/api/envoy/config/route/v3/route.proto#L110

I don't have the same in TRDS at the moment. It should be coming soon in this diff or the upcoming diff.

Comment (htuch) [reply source/common/rds/route_config_update_receiver_impl.cc:19]: FWIW I'm talking about https://github.com/envoyproxy/envoy/blob/b34fcee23881be2e9d5b6e5f04ca6cba02ef4539/api/envoy/config/route/v3/route.proto#L97

Comment (htuch) [reply source/common/rds/route_config_update_receiver_impl.cc:19]: OK, I see. Can the PR title/description/comments be reflected to show this state? I feel it's a bit confusing trying to understand the dynamics here between static vs. dydnamic validation in RDS vs. TRDS.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: done. It makes worried that `ClusterInfoMaps` is concrete class (i.e., not pointer or ref) but we copy here and there. Might need another diff for envoy repo.

File: source/common/rds/route_config_update_receiver_impl.cc
Comment (JuniorHsu) [reply source/common/rds/route_config_update_receiver_impl.cc:19]: Done. Thanks

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (htuch) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:76]: Is there a test for this case?

**Review by htuch — COMMENTED** [review]
LGTM modulo testing question.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:76]: Good catch! Will add one in next commit. Thanks.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: @htuch 

https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/envoy/upstream/cluster_manager.h#L277

`clusters()` return a copy. Hence it doesn't benefits performance for OptRef at the moment. Could we put it for a followup given we should see through the repo, change the signature and make performance improvement?

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (JuniorHsu) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:76]: After I took another look, actually it's a dead code (facepalm)
In http world, we can specified `cluster_header` filed inside the weighted clusters
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/config/route/v3/route_components.proto#L361

Hence it make senses to check either we can validate the cluster or we have a cluster_header.

However, in thrift world. We don't have `cluster_header` for weighted clusters.
From the proto, we can see it's one of the three. We can't get both header and weighted cluster at the moment.
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/config/route/v3/route_components.proto#L361

The `WeightedCluster::ClusterHeader` is propagated from its parent, which doesn't make senses. (note that `ClusterHeader` is one of the interface `RouteEntry` but I don't want to deviate too much at this moment)

To conclude, I remove this dead code given the proto sanitizer will do it for me 
also remove the empty check for the same reason
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/extensions/filters/network/thrift_proxy/v3/route.proto#L117

And add a test to ensure the coverage

**Comment by JuniorHsu** [comment]
@htuch it's ready to have another review with three follow-up items
(a) make `validate_clusters` configurable (I can do it after this lands)
(b) consider cluster manager return a ref
(c) consider remove `ClusterHeader()` from `ThriftProxy::RouteEntry`

thanks

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc
Comment (htuch) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:196]: Yes, not picky here, just thought it looked a bit odd.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by JuniorHsu** [comment]
> ```
> Writing fixes to clang-tidy-fixes.yaml ...
> clang-tidy check failed, potentially fixed by clang-apply-replacements:
> Diagnostics:
> - BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
>   DiagnosticMessage:
>     FileOffset: 298
>     FilePath: ./source/common/common/logger.h
>     Message: '''source/common/common/logger_impl.h'' file not found'
>     Replacements: []
>   DiagnosticName: clang-diagnostic-error
>   Level: Error
>   Ranges:
>   - FileOffset: 298
>     FilePath: ./source/common/common/logger.h
>     Length: 36
> ```
> 
> For the test failure in "envoy-presubmit (check linux_x64 clang_tidy)" , I guess it's unrelated since I didn't touch nor include the logger.

This might need a manual merge.
The clang-tidy error from `Error while processing /source/envoy/rds/config_traits.h.`
which includes context files which include logger

However the clang-tidy can't parse `strip_include_prefix`
https://github.com/envoyproxy/envoy/blob/2be0d7dbe4d66a513bdbbe8a81b06f03d5a69bde/source/common/common/BUILD#L218

**Comment by wrowe** [comment]
The logger was indirectly added by including "envoy/server/factory_context.h", probably without adding the deps for it to envoy/rds/config_traits.h

**Comment by wrowe** [comment]
/wait for deps to be patched up

**Comment by JuniorHsu** [comment]
> The logger was indirectly added by including "envoy/server/factory_context.h", probably without adding the deps for it to envoy/rds/config_traits.h


It's a promising solution and it takes some effort to break the circular reference. Next commit is comming
```
.-> //envoy/router:router_interface (8a9baa446c4cdc335df7924c31f550a82dcac9f2cefefd75763ac169448146ca)
|   //envoy/rds:rds_interface (8a9baa446c4cdc335df7924c31f550a82dcac9f2cefefd75763ac169448146ca)
|   //envoy/server:factory_context_interface (8a9baa446c4cdc335df7924c31f550a82dcac9f2cefefd75763ac169448146ca)
|   //envoy/http:filter_interface (8a9baa446c4cdc335df7924c31f550a82dcac9f2cefefd75763ac169448146ca)
`-- //envoy/router:router_interface (8a9baa446c4cdc335df7924c31f550a82dcac9f2cefefd75763ac169448146ca)
```

**Comment by wrowe** [comment]
Would a reviewer please revisit? It looks that all the comments and feedback have been addressed.

**Comment by JuniorHsu** [comment]
@htuch could you move this forward? I commited @wrowe 's great suggestion to make the test happier. Thanks.

**Review by htuch — APPROVED** [review]
LGTM.

**Pull request merged**
